### ISSUE TITLE
Issue while de-registering a watcher and field level validateOnEmpty issue

### ISSUE THREAD

**Issue by PrashantP25** [ghiscoding/angular-validation#185]
While working on few use cases, we found 2 more issues which we feel would require a fix. 
1) Issue while de-registering a watcher:  In this case when we add 2 validation using addValidator method and then we remove the 2nd validator using removeValidator, then at [validation-service.js line number 494](https://github.com/ghiscoding/angular-validation/blob/master/src/validation-service.js#L494)  it removes 1st validator from the list. So now from the list 1st validator's watcher is de-registered and is also removed from the watchers list. This introduces issue were any changes in 1st validators model value will not trigger validation check. Also validation for 2nd validator will still happen in-case any value of the model of second validator changes. 
   To resolve this issue we have added one method in [common JS](https://github.com/ghiscoding/angular-validation/blob/master/src/validation-common.js) file to remove array from object where value matches. So [line number 494](https://github.com/ghiscoding/angular-validation/blob/master/src/validation-service.js#L494) will be changed from `_watchers.shift();` to `self.commonObj.arrayRemoveObject(_watchers, 'elmName', formElmObj.fieldName);`

**arrayRemoveObject** method definition : 

```js
/** Quick function to remove an object inside an array by it's given field name and value, return and remove the object found or null
     * @param Array sourceArray
     * @param string searchId: search property id
     * @param string searchValue: value to search
     * @return object found from source array or null
     */
    function arrayRemoveObject(sourceArray, searchId, searchValue) {
      if (!!sourceArray) {
        for (var i = 0; i < sourceArray.length; i++) {
          if (sourceArray[i][searchId] === searchValue) {
            var itemToRemove = sourceArray[i];
	    sourceArray.splice(i,1);
	    return itemToRemove;
          }
        }
      }
      return null;
    }
```

2) This issue is when validateOnEmpty is passes with value as true in addValidation method and then with any subsequent method which does not pass this value to addValidation method. validateOnEmpty is accepted as a property during for specific field but it is set globally for all validations. 
   There is one check in attemptToValidate method at [line number 339](https://github.com/ghiscoding/angular-validation/blob/master/src/validation-service.js#L339) where if field is not required and his value is empty, cancel validation and exit out. Here global _validateOnEmpty is checked. We think we should replace it with field level setting as well so if for any field this setting is on execution will not go within this if condition and cancel validation for that field. 
    So the if condition will look like :
    `var isFieldValidateOnEmpty = (self.commonObj.validatorAttrs && self.commonObj.validatorAttrs.validateOnEmpty);`

    `if(!self.commonObj.isFieldRequired() && !(_validateOnEmpty || isFieldValidateOnEmpty) && (value === "" || value === null || typeof value === "undefined"))` 

Can you please accommodate both of these changes in repo?
I have created PR for you to review it. 
[PR 1](https://github.com/ghiscoding/angular-validation/pull/183) : Changes in Common Js file 
[PR 2](https://github.com/ghiscoding/angular-validation/pull/184) : Changes in Validation service Js file 

Thanks you
Prashant

**Comment by ghiscoding** [comment]
That looks ok but I'm quite busy for the next 2 weeks, I'm not sure when I'll have a chance to push a new version

**Comment by PrashantP25** [comment]
That's Ok. Whenever you have time review and accept my PR.
Thanks :)

**Comment by ghiscoding** [comment]
Oh wow I completely forgot about them. I might have some time this week to review. Thanks for the PR

**Comment by ghiscoding** [comment]
@PrashantP25 
Thanks for your patience and the PRs, I've merged everything and release version `1.5.28` on [NPM](https://www.npmjs.com/package/angular-validation-ghiscoding). 

I'll let you close this issue once you validated that it works as intended

**Comment by PrashantP25** [comment]
Thank you so much @ghiscoding for helping me out. 
Have a wonderful day!!!
